File: 24.合唱队.py
"""
动态规划
https://www.bilibili.com/video/BV1XS4y1G7rs/?spm_id_from=autoNext&vd_source=751d83a470a7b7086d96d8b5dd5762d9
"""

nums = int(input())
heights = input().split(" ")
heights = list(map(int, heights))
left = [0 for i in range(nums)]
right = [0 for i in range(nums)]

for i in range(nums):
    j = nums - i - 1
    left[i] = 1
    for ii in range(i):
        if heights[i] > heights[ii]:
            left[i] = max(left[i], left[ii] + 1)
    right[j] = 1
    for jj in range(nums-1, j, -1):
        if heights[j] > heights[jj]:
            right[j] = max(right[j], right[jj] + 1)

min_removal = nums
for i in range(nums):
    removal = nums - (left[i] + right[i] - 1)
    min_removal = min(min_removal, removal)

print(min_removal)

File: 24.合唱队.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int nums;
    cin >> nums;
    int minRemoveal = nums;
    int removeal, height, j;
    vector<int>heights;
    for(int i=0;i<nums;i++)
    {
        cin >> height;
        heights.push_back(height);
    }
    vector<int>left(nums);
    vector<int>right(nums);

    for(int i=0;i<nums;i++)
    {
        j = nums-i-1;
        left[i] = 1;
        for(int ii=0;ii<i;ii++)
        {
            if(heights[i]>heights[ii])
            {
                left[i]= max(left[i], left[ii]+1);
            }
        }
        right[j] = 1;
        for(int jj=nums-1;jj>j;jj--)
        {
           if(heights[j]>heights[jj])
           {
               right[j] = max(right[j], right[jj]+1);
           }
        }

    }
    for(int i=0;i<nums;i++)
    {
        removeal = nums-(left[i]+right[i]-1);
        minRemoveal = min(removeal, minRemoveal);
    }
    cout<<minRemoveal;
    return 0;
}
